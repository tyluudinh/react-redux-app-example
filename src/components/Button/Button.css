.custom-button {
  height: 50px;
  width: auto;
  padding: 10px;
  line-height: 30px;
  border: 0;
  border-radius: 4px;
  text-align: center;
  display: inline-block;

  /* default green button */
  background-color: #009247;
  color: white;
  text-transform: uppercase;
  font-size: 0.8rem;
  cursor: pointer;
  position: relative;

  .custom-button__icon {
    margin-right: 8px;
  }

  .drop-down-container {
    width: 100%;
    height: auto;
    background: white;
    position: absolute;
    top: 50px;
    left: 0;
    right: 0;
    box-shadow:  0 2px 4px 0 rgba(0, 0, 0, 0.12);
    margin-bottom: 12px;
    visibility: hidden;
    opacity: 0;
    transition: opacity 0.3s ease;
  }
  .drop-down-container--active {
    visibility: visible;
    opacity: 1;
  }

  &.no-border-radius {
    border-radius: 0;
  }

  &.min-padding {
    width: 100%;
    @media screen and (min-width: 33.75em) {
      width: auto;
      padding: 10px 50px;
      margin: auto;
    }
  }

  &.full-width {
    width: 100%;
  }

  &:hover,
  &:focus {
    outline: none;
  }

  // Green button 
  &.green {
    border: 1px solid #009247;
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    &:hover,
    &:focus {
      box-shadow: 0 2px 4px rgba(0,0,0,0.25), 0 4px 6px rgba(0,0,0,0.22);
    }
  }

  /* Green reverse buttons */
  &.green-reverse {
    color: #009247;
    background-color: white;
    border: 1px solid #009247;
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    &:hover,
    &:focus {
      background-color: #009247;
      color: white;
      .custom-button__icon{
        fill: white;
      }
    }
    
    .custom-button__icon{
      fill: #009247;
    }
  }

  /* Blue reverse buttons */
  &.blue-reverse {
    color: #00529c;
    background-color: white;
    border: 1px solid #00529c;
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    &:hover,
    &:focus {
      background-color: #00529c;
      color: white;
      .custom-button__icon{
        fill: white;
      }
    }
    .custom-button__icon{
      fill: #00529c;
    }
  }

  /* Red reverse buttons */
  &.red-reverse {
    color: #d0011b;
    background-color: white;
    border: 1px solid #d0011b;
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    &:hover,
    &:focus {
      background-color: #d0011b;
      color: white;
      .custom-button__icon{
        fill: white;
      }
    }
    .custom-button__icon{
      fill: #d0011b;
    }
  }
}