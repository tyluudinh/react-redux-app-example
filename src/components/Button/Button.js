import React, { Component } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";

import "./Button.css";

export default class Button extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dropDownShow: false
    }
    this.toggleDropdown = this.toggleDropdown.bind(this);
  }
  toggleDropdown() {
    this.setState({ dropDownShow: !this.state.dropDownShow });
  }
  render() {
    const {
      type,
      withIcon,
      children,
      fullWidth,
      minPadding,
      noBorderRadius,
      onClick,
      className,
      dropDown,
      ...rest
    } = this.props;

    function renderButtonIcon() {
      if (withIcon) {
        const { name, style, size } = withIcon;
        var Icon = require(`react-icons/lib/fa/${name}`);
        return (
          <Icon
            size={size ? size : null}
            style={style ? style : {}}
            className="custom-button__icon"
          />
        );
      }
      return null;
    }

    return (
      <div
        className={classNames(
          "custom-button",
          {
            "full-width": fullWidth,
            "no-border-radius": noBorderRadius,
            "min-padding": minPadding
          },
          type ? type : "green",
          className
        )}
        {...rest}
        onClick={(dropDown) ? this.toggleDropdown : onClick}
      >
        {renderButtonIcon()}
        {children}
        {(dropDown) ? (
          <div className={classNames(
            'drop-down-container',
            { 'drop-down-container--active': this.state.dropDownShow }
          )}
          onClick={(e) => e.stopPropagation()}>
            {dropDown}
          </div>
        ) : null}
      </div>
    );
  }
}

Button.propTypes = {
  type: PropTypes.string,
  withIcon: PropTypes.object,
  fullWidth: PropTypes.bool,
  noBorderRadius: PropTypes.bool,
  onClick: PropTypes.func,
  className: PropTypes.string,
  dropDown: PropTypes.element,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  minPadding: PropTypes.bool,
};
