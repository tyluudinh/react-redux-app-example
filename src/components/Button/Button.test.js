import React from 'react';

import Button from './Button';

describe('Component Button', () => {
  test('To be rendered correctly!', () => {
    const emptyFn = () => {};

    // Normal button cases
    const normalButton = shallow(<Button onClick={emptyFn}>Confirm</Button>)
    const normalButtonJson = toJson(normalButton);
    expect(normalButtonJson).toMatchSnapshot();

    // Fullwidth button
    const fullwidthButton = shallow(<Button onClick={emptyFn} fullWidth>Confirm</Button>)
    const fullwidthButtonJson = toJson(fullwidthButton);
    expect(fullwidthButtonJson).toMatchSnapshot();

    // specific-type button
    const specificTypeButton = shallow(<Button onClick={emptyFn} type="blue-reverse">Create</Button>)
    const specificTypeButtonJson = toJson(specificTypeButton);
    expect(specificTypeButtonJson).toMatchSnapshot();

    // No border radius button
    const squareBorderButton = shallow(<Button onClick={emptyFn} noBorderRadius>Add more icon</Button>)
    const squareBorderButtonJson = toJson(squareBorderButton);
    expect(squareBorderButtonJson).toMatchSnapshot();

    // button with icon
    const buttonWithIcon = shallow(<Button onClick={emptyFn} withIcon="plus">Add more icon</Button>)
    const buttonWithIconJson = toJson(buttonWithIcon);
    expect(buttonWithIconJson).toMatchSnapshot();

    // button with custom Classname
    const customButton = shallow(<Button onClick={emptyFn} className="custom-button">Custom Button</Button>)
    const customButtonJson = toJson(customButton);
    expect(customButtonJson).toMatchSnapshot();
  })

  test('Click event handler to be called correctly', () => {
    const mockClickfn = jest.fn();
    const button = shallow(<Button onClick={mockClickfn}></Button>);
    button.simulate('click');
    expect(mockClickfn.mock.calls.length).toBe(1);
  })
})
