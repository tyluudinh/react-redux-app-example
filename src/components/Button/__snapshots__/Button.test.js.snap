// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Button To be rendered correctly! 1`] = `
<button
  className="custom-button green"
  onClick={[Function]}
>
  Confirm
</button>
`;

exports[`Component Button To be rendered correctly! 2`] = `
<button
  className="custom-button full-width green"
  onClick={[Function]}
>
  Confirm
</button>
`;

exports[`Component Button To be rendered correctly! 3`] = `
<button
  className="custom-button blue-reverse"
  onClick={[Function]}
>
  Create
</button>
`;

exports[`Component Button To be rendered correctly! 4`] = `
<button
  className="custom-button no-border-radius green"
  onClick={[Function]}
>
  Add more icon
</button>
`;

exports[`Component Button To be rendered correctly! 5`] = `
<button
  className="custom-button green"
  onClick={[Function]}
>
  <FaPlus
    className="custom-button__icon"
  />
  Add more icon
</button>
`;

exports[`Component Button To be rendered correctly! 6`] = `
<button
  className="custom-button green custom-button"
  onClick={[Function]}
>
  Custom Button
</button>
`;
